### PR TITLE
Macros in code expressions.

### PR DESCRIPTION
**Pull request by alex-ozdemir**
Previously, one could not use macros in code expressions.

An example:

(program fun ((t term) (s term)) term
  (and t s))

where `and` is a macro.

There wasn't a very good reason for this.
The only issue is that our type-checking machinery for code
assumed that all application heads were (a) programs or (b) declared
(and thus unbound) symbols.

A macro is just a bound symbol, which we do have type information stored
for, so this PR modifies the procedure for type-checking application
heads to look up that type information.

Macros in code *patterns* are still prohibited.

Credits to @ajreynol  for bringing this up, and I believe to @yoni206 for bringing up something similar a while ago.

### CODE REVIEW

File: tests/tests/code_macro_app.plf [hunk 1]
```diff
@@ -0,0 +1,579 @@
...
+  ((: a term)
+   (: b term)
+   (: a_pf (holds a))
+   (: b_pf (holds b)))
+  (holds (and a b))
+  (and_pf _ _ _ a_pf b_pf))
+
+    
+    
+;(program sc_string_reduction ((t term) (k term)) term
```
Comment (ajreynol) [inline tests/tests/code_macro_app.plf:533]: Let's just delete this.

**Review by ajreynol — CHANGES_REQUESTED** [review]
Thanks for adding this, one minor comment.

File: tests/tests/code_macro_app.plf (hunk 1 above)
Comment (alex-ozdemir) [reply tests/tests/code_macro_app.plf]: Fair enough. I'll do you one better. eb113b5

**Review by ajreynol — APPROVED** [review]
LGTM

**Pull request merged**